slider/beatmap.py

```diff
@@ -1414,31 +1414,150 @@ def ar(self,
 
         return ar
 
-    @lazyval
-    def hit_objects_no_spinners(self):
-        """The hit objects with spinners filtered out.
+    @memoize
+    def hit_objects(self,
+                    spinners=True,
+                    stacking=True,
+                    easy=False,
+                    hard_rock=False,
+                    double_time=False,
+                    half_time=False):
+        """Retrieve hit_objects.
+
+        Parameters
+        ----------
+        spinners : bool, optional
+            If spinners should be included.
+        stacking : bool, optional
+            If stacking should be calculated.
+        easy : bool, optional
+            Calculate stacking with the easy mod enabled.
+        hard_rock : bool, optional
+            Hit_objects will be flipped on the y axis. Calculate stacking with the hard rock mod enabled.
+        half_time : bool, optional
+            Calculate stacking with the half time mod enabled.
+        double_time : bool, optional
+            Calculate stacking with the double time mod enabled.
+
+        Returns
+        -------
+        hit_objects : list[HitObject]
+            The effective AR value.
         """
-        return tuple(e for e in self.hit_objects if not isinstance(e, Spinner))
+
+        hit_objects = self._hit_objects
+
+        if hard_rock:
+            for hit_object in hit_objects:
+                hit_object.position.y = 384 - hit_object.position.y
+
+        if stacking:
+            for i in hit_objects:
+                i.stack_height = 0
```

Comment:
probably best not to use `i` here? Reminiscent of an index instead of an object.